ibis/backends/tests/test_client.py

```diff
@@ -202,3 +203,263 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+            ('department_name', 'string'),
+            ('salary', 'float64'),
+        ]
+    )
+
+    temp_table = 'temp_to_table'
+    temporary = create_temp_table_with_schema(con, temp_table, sch)
+
+    records = pd.DataFrame(
+        {
+            'first_name': ['X', 'Y', 'Z'],
+            'last_name': ['A', 'B', 'C'],
+            'department_name': ['XX', 'YY', 'ZZ'],
+            'salary': [400.0, 500.0, 600.0],
+        }
+    )
+
+    con.insert(temp_table, data_obj=records, if_exists='append')
+    assert len(temporary.execute()) == 3
+    tm.assert_frame_equal(temporary.execute(), records)
+
+
+@pytest.mark.only_on_backends(
+    ['sqlite', 'postgres', 'mysql'],
+    reason="run only if backend is SQLAlchemy based",
+)
+def test_insert_replace_from_dataframe(con):
+    sch = ibis.schema(
+        [
+            ('first_name', 'string'),
+            ('last_name', 'string'),
+            ('department_name', 'string'),
+            ('salary', 'float64'),
+        ]
+    )
+
+    temp_table = 'temp_to_table'
+    temporary = create_temp_table_with_schema(con, temp_table, sch)
+
+    df = pd.DataFrame(
+        {
+            'first_name': ['A', 'B', 'C'],
+            'last_name': ['D', 'E', 'F'],
+            'department_name': ['AA', 'BB', 'CC'],
+            'salary': [100.0, 200.0, 300.0],
+        }
+    )
+    con.load_data(temp_table, df, if_exists='append')
+    assert len(temporary.execute()) == 3
+
+    records = pd.DataFrame(
+        {
+            'first_name': ['X', 'Y', 'Z'],
+            'last_name': ['A', 'B', 'C'],
+            'department_name': ['XX', 'YY', 'ZZ'],
+            'salary': [400.0, 500.0, 600.0],
+        }
+    )
+
+    con.insert(temp_table, data_obj=records, if_exists='replace')
+    assert len(temporary.execute()) == 3
+    tm.assert_frame_equal(temporary.execute(), records)
+
+
+@pytest.mark.only_on_backends(
+    ['sqlite', 'postgres', 'mysql'],
+    reason="run only if backend is SQLAlchemy based",
+)
+def test_insert_fail_from_dataframe(con):
+    sch = ibis.schema(
+        [
+            ('first_name', 'string'),
+            ('last_name', 'string'),
+            ('department_name', 'string'),
+            ('salary', 'float64'),
+        ]
+    )
+
+    temp_table = 'temp_to_table'
+    temporary = create_temp_table_with_schema(con, temp_table, sch)
+
+    df = pd.DataFrame(
+        {
+            'first_name': ['A', 'B', 'C'],
+            'last_name': ['D', 'E', 'F'],
+            'department_name': ['AA', 'BB', 'CC'],
+            'salary': [100.0, 200.0, 300.0],
+        }
+    )
+    con.load_data(temp_table, df, if_exists='append')
+    assert len(temporary.execute()) == 3
+
+    records = pd.DataFrame(
+        {
+            'first_name': ['X', 'Y', 'Z'],
+            'last_name': ['A', 'B', 'C'],
+            'department_name': ['XX', 'YY', 'ZZ'],
+            'salary': [400.0, 500.0, 600.0],
+        }
+    )
+
+    with pytest.raises(ValueError):
```

Comment:
`raises` has a `match` argument that you can use to check that the error message is the one you expect. It's a regex, no need to compare the whole message if it's too long or doesn't make sense. But better to check that this is not failing for some other reason.

Reply:
Sure. One point to note is, `ValueError` is raised by newer versions of pandas (like https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.to_sql.html) but not by older versions (like https://pandas.pydata.org/pandas-docs/version/0.18/generated/pandas.DataFrame.to_sql.html).